Use Isolation library

lontivero opened the pull request:
This PR introduces a new library (Moq) for unit testing. The dependency is for testing only and it helps to isolate components by mocking existing types. 

### Pros

* It makes easier to create mocks 
* It eliminates the needs for creating poor-man mock classes as MockRpcClient and MockTorHttpClient
* It can save time and reduce maintenance cost (reduce code) 
* It allows to do behavioral testing (bad practice imo but anyway)

### Cons

* It is something new to learn
* It is not so trivial to configure mocks with some custom behaviors 
* It could frustrate less experienced devs in some situations
* It could be unnecessary given we have only a few external dependencies: Tor/HttpClient, RPC, FileSystem

There are many many many other isolation libraries and the trends change all the time, that's why I took Moq and no other (and because @kiminuo mentioned it in other PR of mine).

The idea is evaluate it and decide. Ive just used it in a couple of already existing tests to show how it is used (my changes can be drastically improved by sure).

Please vote with your Concept N/ACK.

kiminuo commented:
Concept ACK.

>     * It is something new to learn

This is true. However, mocking is widely used and Moq is one of the main libraries in use in .NET ecosystem. So it is not wasted time because a developer can use the knowledge even elsewhere. There is also https://github.com/Moq/moq4/wiki/Quickstart which is helpful to start. 

>     * It is not so trivial to configure mocks with some custom behaviors

This is true too. However, it is also true that Moq makes some scenarios easier than it is without the library: https://github.com/Moq/moq4/wiki/Quickstart#miscellaneous. 

>     * It could be unnecessary given we have only a few external dependencies: Tor/HttpClient, RPC, FileSystem

Moq can be used to mock virtual methods (not only interfaces) so I think one can mock even other components that are hard-ish to set up or by their nature time-sensitive, etc. which makes them hard to use them in a test.

molnard commented:
This is new for me, but it results in more compact unit tests. Concept ACK

lontivero commented:
I have removed `MockRandom`, `MockTorHttpClient`, `TransactionStoreMock`, `AllTransactionStoreMock` and replaced them by Moq's mocks. There are basically other two candidates: `MockRpcClient` and `HwiProcessBridgeMock` but this PR is already big.

kiminuo commented:
How come that there are not package.deps.json changes? That kind of suggests that the `Moq` dependency is already in the tree .. I don't know.

kiminuo commented on WalletWasabi.Tests/Helpers/ServiceFactory.cs:
```diff
@@ -40,8 +41,8 @@ public static class ServiceFactory
 			var coinsView = new CoinsView(scoins);
-			var transactionStore = new AllTransactionStoreMock(workFolderPath: ".", Network.Main);
-			return new TransactionFactory(Network.Main, keyManager, coinsView, transactionStore, password, allowUnconfirmed);
+			var mockTransactionStore = new Mock<AllTransactionStore>(".", Network.Main);
```
I would also consider using `MockBehavior.Strict`: https://github.com/Moq/moq4/wiki/Quickstart#customizing-mock-behavior

lontivero replied:
No. That would force us to setup the mock. It could be considered to use CallBase instead but it seems this mock is only to fill a gap.

kiminuo commented on WalletWasabi.Tests/UnitTests/Transactions/TransactionStoreTests.cs:
```diff
@@ -79,5 +77,19 @@ public async Task CanDoOperationsAsync()
+			var filePath = Path.Combine(dir, "Transactions.dat");
+			if (File.Exists(filePath))
+			{
+				File.Delete(filePath);
```
Fun fact: `File.Delete` would suffice here. `File.Exists` is not needed actually because: 

> If the file to be deleted does not exist, no exception is thrown.
(https://docs.microsoft.com/en-us/dotnet/api/system.io.file.delete?f=&view=net-5.0)

lontivero replied:
It fails with: `System.IO.DirectoryNotFoundException : Could not find a part of the path`

kiminuo commented on WalletWasabi.Tests/UnitTests/Transactions/TransactionStoreTests.cs:
```diff
@@ -29,9 +29,7 @@ public async Task CanInitializeAsync()
-			await using var txStore = new TransactionStoreMock();
-			var network = Network.Main;
-			await txStore.InitializeAsync(network);
+			var txStore = await CreateTransactionStoreAsync();
```
```suggestion
			await using var txStore = await CreateTransactionStoreAsync();
```

lontivero replied:
I don't get it. `await CreateTransactionStoreAsync()` returns a `TransactionStore` which is not awaitable and does not implement `IDisposable`.

molnard replied:
That is true but it is implementing IAsyncDisposable.

kiminuo commented on WalletWasabi.Tests/UnitTests/Transactions/TransactionStoreTests.cs:
```diff
@@ -12,11 +14,9 @@ public class TransactionStoreTests
-			await using var txStore = new TransactionStoreMock();
-			var network = Network.Main;
-			await txStore.InitializeAsync(network);
+			var txStore = await CreateTransactionStoreAsync();
```
```suggestion
			await using var txStore = await CreateTransactionStoreAsync();
```

kiminuo commented on WalletWasabi.Tests/UnitTests/Crypto/ZeroKnowledge/KnowledgeOfRepTests.cs:
```diff
@@ -29,9 +30,9 @@ public void End2EndVerificationSimple(uint scalarSeed1, uint scalarSeed2)
-			random.GetBytesResults.Add(new byte[32]);
-			var proof = ProofSystemHelpers.Prove(statement, secrets, random);
+			var mockRandom = new Mock<WasabiRandom>();
+			mockRandom.Setup(rnd => rnd.GetBytes(32)).Returns(new byte[32]);
```
This line is very fragile actually. When you change GetBytes(32) -> GetBytes(31), the test still passes.

```csharp
			mockRandom.Setup(rnd => rnd.GetBytes(32)).Returns(new byte[32]).Verifiable();
			var proof = ProofSystemHelpers.Prove(statement, secrets, mockRandom.Object);
			Assert.True(ProofSystemHelpers.Verify(statement, proof));

			mockRandom.Verify();
```

This seems more robust as `mockRandom.Verify();` actually verifies that your mockRandom.Setup was called.

kiminuo replied:
Then your test is a black box and your PR does not make any sense.

lontivero replied:
We have to test black boxes, yes.  This is because we have to verify **what** that the SUT does.  Behavioral testing verify **how** the SUT does it.  These two alternatives have nothing to do with the library, we could do behavioral testing with our PM mocks without any problem. Nevertheless you will not find BT in Wasabi.

Anyway, I switched the mock to strict mode.

kiminuo replied:
> We have to test black boxes, yes.

No. Your test is a black box. Not the tested code. This is simply because when you let anybody read this test, he will rightly assume that `mockRandom.Setup(rnd => rnd.GetBytes(32)).Returns(new byte[32])` actually does something or is needed for some reason.

Is it needed? No. You can delete the line and the test runs just fine. That's very wrong because it is confusing.

So either we will use Moq strict mode everywhere or we can just revert to the old behavior because it was actually better.

nopara73 reviewed (dismissed):
LGTM. Don't get used to adding libs.
